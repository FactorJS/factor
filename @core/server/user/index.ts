export * from "./serverTypes"
export * from "./serverUser"
