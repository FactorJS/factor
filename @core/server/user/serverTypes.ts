import {
  EndpointResponse,
  FullUser,
  PublicUser,
  PrivateUser,
} from "@factor/types"

export const bearerRequired: (keyof UserEndpoint)[] = [
  "currentUser",
  "updateCurrentUser",
  "sendOneTimeCode",
]

export type UserEndpoint = {
  sendOneTimeCode: {
    request: { email: string; userId?: string }
    response: EndpointResponse<boolean>
  }
  updateCurrentUser: {
    request: Partial<FullUser> & { password?: string }
    response: EndpointResponse<FullUser> & { token?: string }
  }
  startNewUser: {
    request: { email: string; fullName?: string }
    response: EndpointResponse<Partial<FullUser>> & {
      token: string
      user: PublicUser
    }
  }
  newVerificationCode: {
    request: { email: string; newAccount?: boolean }
    response: EndpointResponse<{ exists: boolean }>
  }
  verifyAccountEmail: {
    request: { email: string; verificationCode: string }
    response: EndpointResponse<FullUser>
  }
  setPassword: {
    request: { password: string; email: string; verificationCode: string }
    response: EndpointResponse<FullUser> & { token: string }
  }
  resetPassword: {
    request: { email: string }
    response: EndpointResponse<FullUser>
  }
  login: {
    request: { email: string; password: string }
    response: EndpointResponse<FullUser> & {
      token: string
      user: FullUser
      next?: "verify"
    }
  }
  currentUser: {
    request: { token: string }
    response: EndpointResponse<FullUser>
  }
  getPublicUser: {
    request: { email: string } | { userId: string }
    response: EndpointResponse<PublicUser | false>
  }
}

export type UserFetch<U extends keyof UserEndpoint> = {
  request: UserEndpoint[U]["request"]
  response: UserEndpoint[U]["response"]
  endpoint: `/user/${U}`
  method: U
}

export type UserEndpointMethod<U extends keyof UserEndpoint> = (
  args: UserEndpoint[U]["request"],
) => Promise<UserEndpoint[U]["response"]>

export type UserEndpointMethodWithBearer<U extends keyof UserEndpoint> = (
  args: UserEndpoint[U]["request"] & {
    bearer: PrivateUser
    userId: string
  },
) => Promise<UserEndpoint[U]["response"]>

export type EndpointMethods = {
  [K in keyof UserEndpoint]:
    | UserEndpointMethod<K>
    | UserEndpointMethodWithBearer<K>
}
