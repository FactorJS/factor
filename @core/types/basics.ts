export enum CliCommand {
  Start = "start",
  Dev = "dev",
  Build = "build",
}

export const logCategory = {
  debug: { color: "#00ABFF", priority: 5 },
  event: { color: "#0471FF", priority: 10 },
  info: { color: "#00ABFF", priority: 10 },
  notify: { color: "#FF9500", priority: 10 },
  record: { color: "#FF9500", priority: 10 },
  request: { color: "#ff7de5", priority: 10 },
  data: { color: "#FF9500", priority: 10 },
  send: { color: "#00BD0C", priority: 10 },
  service: { color: "#00ffff", priority: 20 },
  command: { color: "#FF9500", priority: 20 },
  build: { color: "#ffdf40", priority: 20 },
  generate: { color: "#00ffa2", priority: 20 },
  success: { color: "#00BD0C", priority: 20 },
  warn: { color: "#ffa500", priority: 30 },
  error: { color: "#FF0000", priority: 40 },
  important: { color: "#00BD0C", priority: 40 },
}
