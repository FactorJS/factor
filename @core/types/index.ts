export * from "./attachment"
export * from "./basics"
export * from "./db"
export * from "./dictionary"
export * from "./email"
export * from "./endpoint"
export * from "./entry"
export * from "./error"
export * from "./geo"
export * from "./ops"
export * from "./post"
export * from "./postTypes"
export * from "./render"
export * from "./roles"
export * from "./server"
export * from "./templates"
export * from "./transform"
export * from "./user"
export * from "./utils"
export * from "./settings"
