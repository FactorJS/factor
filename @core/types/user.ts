import { UserGeolocation } from "./geo"
import { FactorPost, PopulatedPosts, PostActions } from "./post"
import { UserRoles } from "./roles"

export type CurrentUserState = PrivateUser | undefined

enum Gender {
  Male = "male",
  Female = "female",
  Other = "other",
}

export interface DetermineUpdatePermissions {
  bearer: CurrentUserState
  post?: FactorPost
  action: PostActions
  postType: string
  manyPosts?: boolean
}

export type AuthenticationParameters = {
  newAccount?: boolean
  email: string
  password: string
  displayName?: string
  ipAddress?: string
}
/**
 * Publicly accessible user information
 */
export interface PublicUser {
  userId: string
  createdAt: string
  updatedAt: string
  avatar: string
  status: string
  lastSeen?: string | number | Date
  username?: string
  fullName?: string
  email: string
  emailVerified?: boolean
  role?: UserRoles
  profile: Record<string, any>
  settings: Record<string, any>
}
/**
 * Information regarding a user profile (e.g. birthday, cover, tags)
 */
export interface ProfileUser extends PublicUser {
  cover?: PopulatedPosts
  birthday?: Date | string
  gender?: Gender
  about?: string
  tag: string[]
  category: string[]
  site?: string
  github?: string
  githubFollowers?: number
  twitter?: string
  twitterFollowers?: number
  linkedin?: string
  facebook?: string
  workName?: string
  workSeniority?: string
  workRole?: string
  workRoleSub?: string
  workTitle?: string
  workDomain?: string
  bio?: string
  location?: string
}
/**
 * Information including information that should only be available to the
 * current user.
 */
export interface PrivateUser extends PublicUser {
  birthday: string
  phoneNumber?: string
  hashedPassword?: string
  token?: string
  setting?: Record<string, any>
  meta?: Record<string, any>
  verificationCode?: string
  codeExpiresAt?: string | number | Date
  geo?: UserGeolocation
  invitedBy?: string
}
/**
 * All user information
 */
export type FullUser = PrivateUser & ProfileUser

export type AuthCallback = (args: {
  user?: FullUser
  searchBot: boolean
}) => string | undefined | Promise<string | undefined>
