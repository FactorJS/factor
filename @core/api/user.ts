import { PublicUser, FullUser, PrivateUser } from "@factor/types"
import { computed } from "vue"
import { emitEvent } from "./event"
import { clientToken } from "./jwt"
import { logger } from "./logger"
import { getRouter } from "./router"
import { stored, storeItem } from "./store"

export const getJsonUserFields = (): (keyof FullUser)[] => {
  return ["settings", "profile"]
}

export const getEditableUserFields = (): (keyof FullUser)[] => {
  const baseEditable: (keyof FullUser)[] = [
    "fullName",
    "avatar",
    "birthday",
    "cover",
    "gender",
    "about",
  ]

  return baseEditable
}

export const getPublicUserFields = (): (keyof PublicUser)[] => {
  const baseEditable: (keyof PublicUser)[] = [
    "userId",
    "email",
    "emailVerified",
    "role",
    "username",
    "fullName",
    "createdAt",
    "updatedAt",
    "avatar",
    "status",
    "lastSeen",
    "username",
    "profile",
  ]
  return baseEditable
}

export const cacheUser = ({ user }: { user: Partial<FullUser> }): void => {
  if (user && user.userId) storeItem(user.userId, user)
}

/**
 * Information for the currently logged in user
 */
export const currentUser = (): PrivateUser | undefined => {
  return stored("currentUser")
}
/**
 * Active user computed
 */
export const activeUser = computed(() => {
  return currentUser()
})
/**
 * Is the current visitor logged in?
 */
export const isLoggedIn = (): boolean => {
  return currentUser() ? true : false
}
/**
 * Logs out current user and deletes local data
 */
export const deleteCurrentUser = (): void => {
  logger.log({
    level: "info",
    context: "user",
    description: "deleted current user",
  })
  clientToken({ action: "destroy" })
  storeItem("currentUser", undefined)
}

/**
 * Set persistent user info
 */
export const setCurrentUser = (args: {
  user: (PrivateUser & Partial<FullUser>) | undefined
  token?: string
}): void => {
  const { user, token = "" } = args

  if (!user) return deleteCurrentUser()

  storeItem("currentUser", user)
  cacheUser({ user })

  if (token) {
    clientToken({ action: "set", token })
  }
}
/**
 * Logs out the current user
 */
export const logout = async (
  args: { redirect?: string } = {},
): Promise<void> => {
  deleteCurrentUser()
  emitEvent("logout")
  emitEvent("notify", "Successfully logged out.")

  const theCurrentRoute = getRouter().currentRoute.value

  if (
    args.redirect ||
    !theCurrentRoute ||
    theCurrentRoute.matched.some((r) => r.meta.auth)
  ) {
    const { redirect: path = "/" } = args
    await getRouter().push({ path })
  } else {
    emitEvent("resetUi")
  }
}

/**
 * Updates the current locally stored user
 */
export const updateUser = async (
  cb: (
    user: PrivateUser | undefined,
  ) => PrivateUser | undefined | Promise<PrivateUser | undefined>,
): Promise<void> => {
  const user = currentUser()
  const newUser = await cb(user)
  if (newUser) {
    setCurrentUser({ user: newUser })
  }
}
