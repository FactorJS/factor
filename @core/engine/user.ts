/* eslint-disable @typescript-eslint/no-use-before-define */
import { _stop } from "@factor/api/error"
import { runProcessors } from "@factor/api/extend"
import {
  getPublicUserFields,
  getJsonUserFields,
  getEditableUserFields,
} from "@factor/api/user"
import { serverUrl } from "@factor/api/url"
import { validateEmail, snakeCase } from "@factor/api/utils"
import { logger } from "@factor/api/logger"
import { getServerConfig } from "../server/config"
import {
  EndpointResponse,
  FactorTable,
  FullUser,
  PublicUser,
} from "@factor/types"
import bcrypt from "bcrypt"
import dayjs from "dayjs"

import { getDb } from "./db"

import { createClientToken, decodeClientToken } from "./jwt"

import { EndpointMethodOptions, Endpoint, EndpointMeta } from "./endpoint"

import { Query } from "./query"

/**
 * A random 6 digit number, ideal for verification code
 */
const getSixDigitRandom = (): string => {
  return Math.random().toString().slice(2, 8)
}
/**
 * Create a md5 of password
 * @remarks
 * 6 salt rounds - https://security.stackexchange.com/a/83382
 */
const hashPassword = async (password: string): Promise<string> => {
  return await bcrypt.hash(password, 6)
}
const comparePassword = async (
  password: string,
  hashedPassword: string,
): Promise<boolean> => {
  return await bcrypt.compare(password, hashedPassword)
}

/**
 * Add user data about the current user
 */
const processUser = async (
  user: FullUser,
  meta?: { private: boolean },
): Promise<FullUser> => {
  meta = meta || { private: true }
  const config = getServerConfig()

  const processors = config?.user?.processors ?? []

  if (processors && processors.length > 0) {
    const result = await runProcessors<FullUser>(processors, user, meta)
    return result
  } else {
    return user
  }
}

/**
 * Verify a new user email is valid and unique
 */
export const verifyNewEmail = async (email?: string): Promise<true> => {
  if (!email) throw _stop({ message: "email is required" })
  if (!validateEmail(email)) throw _stop({ message: "email failed validation" })

  const { data: exists } = await Queries.ManageUser.serve(
    {
      _action: "getPublic",
      email,
    },
    {},
  )

  if (exists) {
    throw _stop({ message: "email already exists", data: { email } })
  }

  return true
}

class QueryManageUser extends Query {
  async run(
    params:
      | { _action: "create"; fields: Partial<FullUser> & { email: string } }
      | {
          _action: "update"
          fields: Partial<FullUser>
          email?: string
          userId?: string
        }
      | {
          _action: "getPrivate" | "getPublic"
          email?: string
          userId?: string
          select?: (keyof FullUser)[] | ["*"]
        },
    _meta?: EndpointMeta,
  ): Promise<EndpointResponse<FullUser>> {
    const { _action } = params

    const db = await this.getDb()

    let user: FullUser | undefined

    const publicFields = getPublicUserFields()

    if (_action == "getPrivate" || _action == "getPublic") {
      const { userId, email } = params
      const where = userId ? { userId } : { email }

      const returnFields = _action == "getPrivate" ? ["*"] : publicFields

      user = await db
        .select(...returnFields)
        .from(FactorTable.User)
        .where(where)
        .first<FullUser>()

      if (user && _action == "getPrivate") {
        user = await processUser(user)
      }
    } else if (_action == "update") {
      const { userId, email, fields } = params
      if (!fields || (!userId && !email)) {
        throw this.stop("missing required fields")
      }

      const where = userId ? { userId } : { email }

      ;[user] = await db(FactorTable.User)
        .update(fields)
        .where(where)
        .returning<FullUser[]>("*")

      if (!user) throw this.stop({ message: "user not found", data: where })

      user = await processUser(user)
    } else if (_action == "create") {
      const { fields } = params

      if (!fields) throw this.stop("fields required")

      const { email } = fields

      if (!email) throw this.stop("email required")

      await verifyNewEmail(email)
      ;[user] = await db
        .insert({
          ...fields,
          verificationCode: getSixDigitRandom(),
          codeExpiresAt: dayjs().add(7, "day").toISOString(),
        })
        .into(FactorTable.User)
        .returning<FullUser[]>("*")

      if (!user) throw this.stop("problem creating user")
    }

    return { status: "success", data: user }
  }
}

class QueryCurrentUser extends Query {
  async run(params: { token: string }): Promise<EndpointResponse<FullUser>> {
    const { token } = params

    if (!token) throw this.stop({ message: "auth info not sent (token)" })

    const { email } = decodeClientToken(token)

    const { data: user } = await Queries.ManageUser.serve(
      {
        _action: "update",
        email,
        fields: { lastSeen: dayjs().toISOString() },
      },
      undefined,
    )

    return { status: "success", data: user }
  }
}

/**
 * Send a verification email with code
 */
export const sendVerificationEmail = async (args: {
  email: string
  code: string | number
}): Promise<void> => {
  const { sendEmail } = await import("./email")
  const { code, email } = args
  await sendEmail({
    subject: `${code} is your verification code`,
    text: `This email is to verify your account using a one-time code.\n\n Your code is: **${code}**`,
    to: email,
  })
}

/**
 * Create a verification code, save it to user and email the user with the code
 */
export const sendOneTimeCode = async (params: {
  email: string
}): Promise<EndpointResponse<boolean>> => {
  const { email } = params

  const code = getSixDigitRandom()
  const fields = {
    verificationCode: code,
    codeExpiresAt: dayjs().add(1, "day").toISOString(),
  }

  await Queries.ManageUser.serve(
    { _action: "update", email, fields },
    undefined,
  )

  await sendVerificationEmail({ email, code })

  return { status: "success", data: true }
}
class QuerySendOneTimeCode extends Query {
  async run(params: { email: string }): Promise<EndpointResponse<boolean>> {
    if (!this.qu) throw new Error("no knex")

    return sendOneTimeCode(params)
  }
}

/**
 * Verify if a passed code matches the stored one
 */
export const verifyCode = async (args: {
  email?: string
  userId?: string
  verificationCode: string
}): Promise<true> => {
  const { email, verificationCode, userId } = args

  if (!verificationCode) {
    throw _stop({ message: `no code provided` })
  }
  if (!email && !userId) {
    throw _stop({ message: `need email or userId` })
  }

  const where = userId ? { userId } : { email }

  const db = await getDb()
  const r = await db
    .select<{ verificationCode: string; codeExpiresAt: string }[]>([
      "verificationCode",
      "codeExpiresAt",
    ])
    .from(FactorTable.User)
    .where(where)

  const codeDetails = r && r.length > 0 ? r[0] : undefined

  // allow short circuit in development
  if (process.env.NODE_ENV == "development" && verificationCode == "123456") {
    return true
  }

  if (!codeDetails || codeDetails.verificationCode !== verificationCode) {
    throw _stop({ message: `verification code is not a match` })
  } else if (
    !codeDetails.codeExpiresAt ||
    dayjs().isAfter(codeDetails.codeExpiresAt)
  ) {
    throw _stop({ message: `verification code is expired` })
  }

  return true
}

/**
 * Updates the current user with new info
 * Detecting if auth fields have changed and verifying account code
 */
class QueryUpdateCurrentUser extends Query {
  async run(
    params: Partial<FullUser> & { password?: string },
    meta: EndpointMeta,
  ): Promise<EndpointResponse<FullUser> & { token?: string }> {
    if (!this.qu) throw new Error("no knex")
    if (!meta.bearer) throw this.stop({ message: "must be logged in" })

    const { email, userId, password, verificationCode } = params
    const { bearer } = meta
    let fields: Partial<FullUser> = {}

    for (const f of getEditableUserFields()) {
      if (typeof params[f] != "undefined") {
        fields = { ...fields, [f]: params[f] }
      }
    }

    for (const f of getJsonUserFields()) {
      if (typeof params[f] != "undefined") {
        const jsn = JSON.stringify(params[f])

        const setter = this.qu.raw(
          `coalesce(${snakeCase(f)}::jsonb, '{}'::jsonb) || ?::jsonb`,
          jsn,
        )

        fields = { ...fields, [f]: setter }
      }
    }

    if (password) {
      const hashedPassword = await hashPassword(password)

      const { data: dbUser } = await Queries.ManageUser.serve(
        {
          _action: "getPrivate",
          userId: bearer.userId,
        },
        meta,
      )

      if (!dbUser) throw this.stop({ message: "couldn't find user" })

      const correctPassword = await comparePassword(
        password,
        dbUser.hashedPassword ?? "",
      )

      if (verificationCode) {
        await verifyCode({ email, userId, verificationCode })
      }

      if (!correctPassword && !verificationCode) {
        throw this.stop({ message: "account verification required" })
      }

      if (email && email != dbUser.email) {
        await verifyNewEmail(email)

        fields.email = email
      } else {
        fields.hashedPassword = hashedPassword
      }
    }

    let user, token
    if (Object.keys(fields).length > 0) {
      const response = await Queries.ManageUser.serve(
        {
          _action: "update",
          userId: bearer.userId,
          fields,
        },
        meta,
      )

      user = response.data

      if (!user) throw this.stop("problem updating user")

      // if email or password were changed, create new token
      token = password ? createClientToken(user) : undefined
    }

    return {
      status: "success",
      data: user,
      message: "changes saved",
      token,
      user,
    }
  }
}

class QuerySetPassword extends Query {
  async run(
    params: {
      password: string
      email: string
      verificationCode: string
    },
    meta: EndpointMeta,
  ): Promise<EndpointResponse<FullUser> & { token: string }> {
    if (!this.qu) throw new Error("no knex")
    if (!meta.bearer) throw this.stop({ message: "must be logged in" })

    const { password, email: emailArg, verificationCode } = params

    const email = emailArg ?? meta.bearer.email
    if (!email) {
      throw _stop({ message: `email must be verified to set new password` })
    }
    // code verification is needed because on password reset the user is logged out
    await verifyCode({ email, verificationCode })
    const hashedPassword = await hashPassword(password)

    const { data: user } = await Queries.ManageUser.serve(
      {
        _action: "update",
        email,
        fields: { hashedPassword },
      },
      meta,
    )

    if (!user) throw this.stop("problem updating user")

    user.hashedPassword = hashedPassword

    return {
      status: "success",
      data: user,
      message: "new password created",
      token: createClientToken(user),
      user,
    }
  }
}

class QueryVerifyAccountEmail extends Query {
  async run(params: {
    email: string
    verificationCode: string
  }): Promise<EndpointResponse<FullUser>> {
    if (!this.qu) throw new Error("no knex")

    const { email, verificationCode } = params

    if (!email) throw _stop({ message: "email is required" })
    if (!verificationCode) throw _stop({ message: "confirm code is required" })

    await verifyCode({ email, verificationCode })

    const { data: user } = await Queries.ManageUser.serve(
      {
        _action: "update",
        email,
        fields: { emailVerified: true },
      },
      undefined,
    )

    if (!user) throw this.stop("problem updating user")

    // send it back for convenience
    user.verificationCode = verificationCode

    const config = getServerConfig()

    const onVerified = config?.user?.onVerified

    if (onVerified) {
      Promise.resolve(onVerified(user)).catch((error) => console.error(error))
    }

    logger.log({
      level: "info",
      context: "QueryVerifyAccountEmail",
      description: `user verified ${email}`,
    })

    return {
      status: "success",
      data: user,
      message: "verification successful",
      token: createClientToken(user),
    }
  }
}

class QueryResetPassword extends Query {
  async run({ email }: { email: string }): Promise<EndpointResponse<FullUser>> {
    if (!this.qu) throw new Error("no knex")

    if (!email) throw this.stop({ message: "email is required" })

    await sendOneTimeCode({ email })

    return {
      status: "success",
      message: "verification code sent",
    }
  }
}

class QueryStartNewUser extends Query {
  async run(params: { email: string; fullName?: string }): Promise<
    EndpointResponse<Partial<FullUser>> & {
      token: string
      user: PublicUser
    }
  > {
    if (!this.qu) throw new Error("no knex")

    const { email, fullName } = params
    const { data: user } = await Queries.ManageUser.serve(
      {
        _action: "create",
        fields: { email, fullName },
      },
      undefined,
    )

    if (!user) throw this.stop("problem creating user")

    await sendOneTimeCode({ email: user.email })

    logger.log({
      level: "info",
      context: "QueryStartNewUser",
      description: `user started ${email}:${fullName ?? "(no name)"}`,
    })

    return {
      status: "success",
      data: { userId: user.userId, fullName: user.fullName, email: user.email },
      message: "verification code sent",
      user,
      token: createClientToken(user),
    }
  }
}

class QueryLogin extends Query {
  async run(
    params: { email: string; password: string },
    _meta: EndpointMeta,
  ): Promise<
    EndpointResponse<FullUser> & {
      token: string
      user: FullUser
      next?: "verify"
    }
  > {
    const { email, password } = params
    let { data: user } = await Queries.ManageUser.serve(
      {
        _action: "getPrivate",
        email,
      },
      _meta,
    )

    if (!user) throw this.stop({ message: "user does not exist" })

    const correctPassword = await comparePassword(
      password,
      user.hashedPassword ?? "",
    )

    if (!correctPassword) {
      throw this.stop({ message: "password is incorrect" })
    }

    delete user.hashedPassword

    user = await processUser(user)

    const token = createClientToken(user)

    if (!user.emailVerified) {
      await sendOneTimeCode({ email: user.email })
      return {
        status: "success",
        data: user,
        message: "verification email sent",
        next: "verify",
        token,
        user,
      }
    }

    logger.log({
      level: "info",
      context: "QueryLogin",
      description: `user logged in ${email}`,
    })

    return {
      status: "success",
      data: user,
      message: "successfully logged in",
      token,
      user,
    }
  }
}

class QueryNewVerificationCode extends Query {
  async run(
    params: {
      email: string
      newAccount?: boolean
    },
    meta: EndpointMeta,
  ): Promise<EndpointResponse<{ exists: boolean }>> {
    if (!this.qu) throw new Error("no knex")

    const { email, newAccount } = params

    let { data: existingUser } = await Queries.ManageUser.serve(
      {
        _action: "getPrivate",
        email,
      },
      meta,
    )

    const exists = existingUser ? true : false

    if (newAccount && exists) {
      throw this.stop({ message: "email exists", data: { exists } })
    } else if (!existingUser) {
      const { data: createdUser } = await Queries.ManageUser.serve(
        {
          _action: "create",
          fields: { email },
        },
        meta,
      )

      existingUser = createdUser

      logger.log({
        level: "info",
        context: "user",
        description: `user created ${email}`,
      })
    }

    if (!existingUser) throw this.stop("no user")

    await sendOneTimeCode({ email: existingUser.email })

    return {
      status: "success",
      data: { exists },
      message: "verification code sent",
    }
  }
}

export class UserMethod<T extends Query> extends Endpoint<T> {
  constructor(options: EndpointMethodOptions<T>) {
    super({ baseURL: serverUrl(), basePath: "/user", ...options })
  }
}

export const Queries = {
  Login: new QueryLogin(),
  NewVerificationCode: new QueryNewVerificationCode(),
  SetPassword: new QuerySetPassword(),
  ResetPassword: new QueryResetPassword(),
  UpdateCurrentUser: new QueryUpdateCurrentUser(),
  SendOneTimeCode: new QuerySendOneTimeCode(),
  VerifyAccountEmail: new QueryVerifyAccountEmail(),
  StartNewUser: new QueryStartNewUser(),
  CurrentUser: new QueryCurrentUser(),
  ManageUser: new QueryManageUser(),
}

type EndpointMap = {
  [P in keyof typeof Queries]: UserMethod<typeof Queries[P]>
}

export const getEndpointsMap = (): EndpointMap => {
  return Object.fromEntries(
    Object.entries(Queries).map(([key, query]) => {
      return [key, new UserMethod({ key, queryHandler: query })]
    }),
  ) as EndpointMap
}

export const endpointsMap = getEndpointsMap()
export const endpointsList = Object.values(endpointsMap)
