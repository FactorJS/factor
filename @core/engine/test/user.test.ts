/* eslint-disable @typescript-eslint/no-unsafe-call */
import { expect, it, vi, describe } from "vitest"
import { Queries } from "../user"
import { FullUser } from "@factor/types"
import * as ep from "../user"
import bcrypt from "bcrypt"

vi.mock("../serverEmail", async () => {
  const actual = (await vi.importActual("../serverEmail")) as Record<
    string,
    any
  >
  return {
    ...actual,
    getEmailSMTPService: vi.fn(() => {
      return { sendEmail: vi.fn() }
    }),
  }
})

let user: FullUser
const key = Math.random().toString().slice(2, 8)

describe("user tests", () => {
  it("creates user", async () => {
    const response = await Queries.ManageUser.serve(
      {
        _action: "create",
        fields: { email: `arpowers+${key}@gmail.com`, fullName: "test" },
      },
      undefined,
    )

    if (!response.data) throw new Error("problem creating user")

    user = response.data

    expect(user?.userId).toBeTruthy()
    expect(user?.fullName).toBe("test")
    expect(user?.verificationCode).toBeTruthy()
  })

  it("verifies account email", async () => {
    const response = await ep.Queries.VerifyAccountEmail.serve(
      {
        email: user.email,
        verificationCode: user.verificationCode ?? "",
      },
      undefined,
    )

    expect(response.data).toBeTruthy()
    expect(response.status).toBe("success")
    expect(response.message).toBe("verification successful")

    user = response.data as FullUser

    expect(user?.emailVerified).toBeTruthy()
    expect(user?.verificationCode).toBeTruthy()
  })

  it("sets password", async () => {
    const response = await ep.Queries.SetPassword.serve(
      {
        email: user.email,
        verificationCode: user.verificationCode ?? "",
        password: "test",
      },
      { bearer: user },
    )
    expect(response.message).toContain("password created")
    user = response.data as FullUser

    expect(bcrypt.compare("test", user?.hashedPassword ?? "")).toBeTruthy()
    expect(user?.token).toMatchSnapshot()
  })

  it("resets password", async () => {
    const response = await ep.Queries.ResetPassword.serve(
      {
        email: user.email,
      },
      undefined,
    )

    expect(response.status).toBe("success")

    if (!response.internal) throw new Error("code required")

    const response2 = await ep.Queries.SetPassword.serve(
      {
        email: user.email,
        verificationCode: response.internal,
        password: "test",
      },
      { bearer: user },
    )

    expect(response2.status).toBe("success")
  })

  it("updates the user", async () => {
    const response = await ep.Queries.ManageUser.serve(
      {
        _action: "update",
        email: user.email,
        fields: {
          fullName: "testUpdate",
          facebook: "https://www.facebook.com/apowers",
        },
      },
      { bearer: user },
    )

    expect(response.status).toBe("success")
    expect(response.data?.fullName).toBe("testUpdate")
    expect(response.data?.facebook).toBe("https://www.facebook.com/apowers")
  })
})
