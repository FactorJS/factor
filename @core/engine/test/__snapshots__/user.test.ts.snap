// Vitest Snapshot v1

exports[`setPassword 1`] = `undefined`;

exports[`user tests > setPassword 1`] = `undefined`;

exports[`user tests > sets password 1`] = `undefined`;
